### PR TITLE
Implementation of the ConvKB Score Function

### PR DESCRIPTION
**Pull request by JWG-1**
Implementation of the ConvKB Scorefunction as in the paper: https://aclanthology.org/N18-2053.pdf

### CODE REVIEW

File: Readme.md [hunk 1]
```diff
@@ -70,6 +70,9 @@ Given node and relation embeddings, CompGCN performs a composition operation φ(
   # CompGCN (Composition: Circular Correlation)
   python run.py -score_func conve -opn corr
   
+  ##### CompGCN with ConvKB ScoreFunction
+  python run.py -score_func convkd -init_dim 100 -num_filt 3
```
Comment (RaphaW) [inline Readme.md]: Here's still the old name written

File: model/models.py [hunk 2]
```diff
@@ -102,7 +102,7 @@ def __init__(self, edge_index, edge_type, params=None):
 		self.fc			= torch.nn.Linear(self.flat_sz, self.p.embed_dim)
 
 	def concat(self, e1_embed, rel_embed):
-		e1_embed	= e1_embed. view(-1, 1, self.p.embed_dim)
+		e1_embed	= e1_embed.view(-1, 1, self.p.embed_dim)
```
Comment (RaphaW) [inline model/models.py]: As Max said before please do not reformat the already existing code. This makes it much harder to review your changes

File: model/models.py [hunk 3]
```diff
@@ -128,3 +128,35 @@ def forward(self, sub, rel):
 
 		score = torch.sigmoid(x)
 		return score
+
```
Comment (RaphaW) [inline model/models.py:131]: Could you add some docu on top of the class.
Something along the lines of how it works, so others also understand it without all too much work

File: model/models.py [hunk 4]
```diff
@@ -128,3 +128,35 @@ def forward(self, sub, rel):
...
+		self.conv2d1 = torch.nn.Conv2d(1, out_channels=self.p.num_filt, kernel_size=(1, 3), stride=1)
+		self.fake_fc_with_conv = torch.nn.Conv2d(self.p.num_filt,out_channels= 1 , kernel_size=(self.p.embed_dim, 1), stride=self.p.embed_dim,bias=False)
+
+		#Prints the number of parameter in the network to the console
+		total_params = sum(
+			param.numel() for param in self.parameters()
+		)
+		print("Total Parameter: " + str(total_params))
+
+	def forward(self, sub,rel):
```
Comment (RaphaW) [inline model/models.py:153]: could you please add some docstring explaining this function?

File: model/models.py [hunk 5]
```diff
@@ -128,3 +128,35 @@ def forward(self, sub, rel):
...
+		all_ent = all_ent.unsqueeze(0).expand(l,-1,-1).reshape((l,1,self.p.num_ent * self.p.embed_dim))
+													#size of tensor after manipulation, bs= batch_size
+		x = torch.cat([sub_emb_repeat, rel_emb_repeat, all_ent], 1) 	#bs x 3 x num_ent * embed_dim
+		x = x.transpose(1, 2) 						#bs x num_ent * embed_dim x 3
+		x = x.unsqueeze(1)  						#bs x 1 x num_ent * embed_dim x 3
+		conv_out = self.conv2d1(x) 					#bs x num_filt x num_ent * embed_dim x 1
+		conv_out = self.relu(conv_out)					#bs x num_filt x num_ent * embed_dim x 1
+		score = self.fake_fc_with_conv(conv_out) 			#bs x 1 x num_ent x 1
+		score = score.view(l,self.p.num_ent) 				#bs x num_ent
+		return torch.sigmoid(score) 					#bs x num_ent
```
Comment (RaphaW) [inline model/models.py]: What are all the comments on the right here ? If they should explain something, maybe do not use abbreviations

File: Readme.md (hunk 1 above)
Comment (JWG-1) [reply Readme.md]: fixed

File: model/models.py (hunk 2 above)
Comment (JWG-1) [reply model/models.py]: fixed, but i looks weird to write it like that with a space between

File: model/models.py (hunk 4 above)
Comment (JWG-1) [reply model/models.py:153]: added documentation

File: model/models.py (hunk 5 above)
Comment (JWG-1) [reply model/models.py]: they are showing the size of the tensor after each manipulation to help with understanding what is going on, bs is defined num_filt = Number of filters
num_ent = Number of entities
embed_dim = embedding dimension

File: model/models.py (hunk 3 above)
Comment (JWG-1) [reply model/models.py:131]: added the documentation to the function forward

Comment (JWG-1) [reply model/models.py:131]: also added some docu. at the top

**Comment by 1Maxnet1** [comment]
Someone accidentally pushed the .idea directory to this. Could you please remove it again?

File: model/models.py (hunk 4 above)
Comment (1Maxnet1) [reply model/models.py:153]: Instead of plenty of inline comments, could you shift the explanation to the docstring?

**Review by 1Maxnet1 — COMMENTED** [review]
Thanks for incooperating all the feedback! Here is my renewed review. Everything works as expected now.

File: model/models.py (hunk 4 above)
Comment (JWG-1) [reply model/models.py:153]: ok

**Pull request merged**